api/traces/v1/http.go

```diff
@@ -0,0 +1,291 @@
+func WithWriteMiddleware(m func(http.Handler) http.Handler) HandlerOption {
+	return func(h *handlerConfiguration) {
+		h.writeMiddlewares = append(h.writeMiddlewares, m)
+	}
+}
+
+type handlerInstrumenter interface {
+	NewHandler(labels prometheus.Labels, handler http.Handler) http.HandlerFunc
+}
+
+type nopInstrumentHandler struct{}
+
+func (n nopInstrumentHandler) NewHandler(labels prometheus.Labels, handler http.Handler) http.HandlerFunc {
+	return handler.ServeHTTP
+}
+
+// NewV2APIHandler creates a trace query handler for Jaeger V2 HTTP queries
+func NewV2APIHandler(read *url.URL, opts ...HandlerOption) http.Handler {
+	c := &handlerConfiguration{
+		logger:     log.NewNopLogger(),
+		registry:   prometheus.NewRegistry(),
+		instrument: nopInstrumentHandler{},
+	}
+
+	for _, o := range opts {
+		o(c)
+	}
+
+	r := chi.NewRouter()
+
+	var proxyRead http.Handler
+	{
+		level.Debug(c.logger).Log("msg", "Configuring upstream Jaeger query v2", "queryv2", read)
+		middlewares := proxy.Middlewares(
+			proxy.MiddlewareSetUpstream(read),
+			proxy.MiddlewareSetPrefixHeader(),
+			proxy.MiddlewareLogger(c.logger),
+			proxy.MiddlewareMetrics(c.registry, prometheus.Labels{"proxy": "tracesv1-read"}),
+		)
+
+		t := &http.Transport{
+			DialContext: (&net.Dialer{
+				Timeout: ReadTimeout,
+			}).DialContext,
+		}
+
+		proxyRead = &httputil.ReverseProxy{
+			Director:     middlewares,
+			ErrorLog:     proxy.Logger(c.logger),
+			Transport:    otelhttp.NewTransport(t),
+			ErrorHandler: func(rw http.ResponseWriter, r *http.Request, e error) {},
+		}
+	}
+	r.Group(func(r chi.Router) {
+		r.Use(c.readMiddlewares...)
+		r.Handle("/*", c.instrument.NewHandler(
+			prometheus.Labels{"group": "tracesv1api", "handler": "api"},
+			proxyRead))
+	})
+
+	return r
+}
+
+// NewUIStaticHandler creates a trace handler for Jaeger UI static assets (.js and .css)
+func NewUIStaticHandler(read *url.URL, opts ...HandlerOption) http.Handler {
+	c := &handlerConfiguration{
+		logger:     log.NewNopLogger(),
+		registry:   prometheus.NewRegistry(),
+		instrument: nopInstrumentHandler{},
+	}
+
+	for _, o := range opts {
+		o(c)
+	}
+
+	r := chi.NewRouter()
+
+	var proxyRead http.Handler
+	{
+		level.Debug(c.logger).Log("msg", "Configuring upstream Jaeger UI static", "ui", read)
+		middlewares := proxy.Middlewares(
+			proxy.MiddlewareSetUpstream(read),
+			proxy.MiddlewareSetPrefixHeader(),
+			proxy.MiddlewareLogger(c.logger),
+			proxy.MiddlewareMetrics(c.registry, prometheus.Labels{"proxy": "tracesv1-ui-static"}),
+		)
+
+		t := &http.Transport{
+			DialContext: (&net.Dialer{
+				Timeout: ReadTimeout,
+			}).DialContext,
+		}
+
+		proxyRead = &httputil.ReverseProxy{
+			Director:  middlewares,
+			ErrorLog:  proxy.Logger(c.logger),
+			Transport: otelhttp.NewTransport(t),
+		}
+	}
+	r.Group(func(r chi.Router) {
+		r.Use(c.readMiddlewares...)
+		r.Handle("/*", c.instrument.NewHandler(
+			prometheus.Labels{"group": "tracesv1ui", "handler": "search"},
```

Comment:
since `tracesv1ui` and `search` are reused in `NewUIHandler` maybe they could be moved to consts?

Reply:
I am trying to follow the style of https://github.com/observatorium/api/blob/main/api/metrics/v1/http.go#L167 where the labels are explicit.  I like that style; it is very readable.  The compiler will optimize the literals together.

Reply:
sounds good :+1: